### PR TITLE
Feature/ml word2vec example

### PR DESCRIPTION
**Pull request by baykaner**

### CODE REVIEW

File: libs/math/include/math/free_functions/matrix_operations/matrix_operations.hpp [hunk 1]
```diff
@@ -543,26 +543,27 @@ void PeakToPeak(ArrayType arr)
  * @param array array to search for maximum value
  * @return
  */
-template <typename T, typename C>
-void ArgMax(ShapelessArray<T, C> const &array, T &ret)
+template <typename ArrayType>
+void ArgMax(ArrayType const &array, typename ArrayType::Type &ret)
```
Comment (pierre-wilmot-fetch-ai) [inline libs/math/include/math/free_functions/matrix_operations/matrix_operations.hpp]: ret should probably be of type ArrayType::SizeType

File: libs/math/include/math/free_functions/matrix_operations/matrix_operations.hpp [hunk 2]
```diff
@@ -543,26 +543,27 @@ void PeakToPeak(ArrayType arr)
...
 {
-  ret          = 0;
-  T cur_maxval = std::numeric_limits<T>::lowest();
+  ret                                 = 0;
+  typename ArrayType::Type cur_maxval = std::numeric_limits<typename ArrayType::Type>::lowest();
 
   // just using ret as a free variable to store the current maxval for the loop here
   for (std::size_t i = 0; i < array.size(); ++i)
```
Comment (pierre-wilmot-fetch-ai) [inline libs/math/include/math/free_functions/matrix_operations/matrix_operations.hpp]: Same for i

File: libs/math/include/math/free_functions/matrix_operations/matrix_operations.hpp [hunk 3]
```diff
@@ -543,26 +543,27 @@ void PeakToPeak(ArrayType arr)
...
   {
     if (cur_maxval < array[i])
     {
-      ret = static_cast<T>(i);
+      ret = static_cast<typename ArrayType::Type>(i);
```
Comment (pierre-wilmot-fetch-ai) [inline libs/math/include/math/free_functions/matrix_operations/matrix_operations.hpp]: And no need for cast anymore

File: libs/math/include/math/free_functions/ml/loss_functions/scaled_cross_entropy.hpp [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+  typename ArrayType::Type plogx = typename ArrayType::Type(0);
+  for (std::size_t i = 0; i < x.shape()[0]; ++i)
+  {
+    for (std::size_t j = 0; j < x.shape()[1]; ++j)
+    {
+      if (y.At({i, j}) == typename ArrayType::Type(1))
+      {
+        typename ArrayType::Type tmp2 = x.At({i, j});
+        typename ArrayType::Type tmp  = Log(tmp2);
+        fetch::math::Add(plogx, fetch::math::Multiply(typename ArrayType::Type(-1), tmp), plogx);
```
Comment (pierre-wilmot-fetch-ai) [inline libs/math/include/math/free_functions/ml/loss_functions/scaled_cross_entropy.hpp]: You can probably break out of the loop once you've found the 1

File: libs/math/include/math/free_functions/ml/loss_functions/softmax_cross_entropy.hpp [hunk 5]
```diff
@@ -48,15 +50,19 @@ ArrayType SoftmaxCrossEntropyLoss(ArrayType const &x, ArrayType const &y)
   // (i.e. x is the output
   //  of softmax layer)
 
-  auto      gt = ArgMax(y, 1);
+  //  auto      gt = ArgMax(y, 1);
```
Comment (pierre-wilmot-fetch-ai) [inline libs/math/include/math/free_functions/ml/loss_functions/softmax_cross_entropy.hpp]: Debug code

File: libs/math/include/math/free_functions/ml/loss_functions/softmax_cross_entropy.hpp [hunk 6]
```diff
@@ -48,15 +50,19 @@ ArrayType SoftmaxCrossEntropyLoss(ArrayType const &x, ArrayType const &y)
...
   log_likelihood[0] = 0;
 
   for (typename ArrayType::SizeType idx = 0; idx < n_examples; ++idx)
   {
-    sce_x.Set(idx, static_cast<typename ArrayType::SizeType>(gt[idx]),
-              std::log(sce_x.At(idx, static_cast<typename ArrayType::SizeType>(gt[idx]))));
-    log_likelihood[0] -= sce_x.At(idx, static_cast<typename ArrayType::SizeType>(gt[idx]));
+    //    sce_x.Set(idx, static_cast<typename ArrayType::SizeType>(gt[idx]), std::log(sce_x.At(idx,
+    //    static_cast<typename ArrayType::SizeType>(gt[idx])))); log_likelihood[0] -= sce_x.At(idx,
+    //    static_cast<typename ArrayType::SizeType>(gt[idx]));
```
Comment (pierre-wilmot-fetch-ai) [inline libs/math/include/math/free_functions/ml/loss_functions/softmax_cross_entropy.hpp]: Debug code

File: libs/math/include/math/free_functions/standard_functions/abs.hpp [hunk 7]
```diff
@@ -29,26 +29,61 @@ namespace fetch {
...
+  ret.data().in_parallel().Apply(kernel, x.data());
+}
+
+template <typename ArrayType>
+fetch::math::meta::IfIsNonBlasArray<ArrayType, void> Abs(ArrayType &x, ArrayType &ret)
+{
+  assert(x.size() == ret.size());
+  for (std::size_t j = 0; j < ret.size(); ++j)
+  {
+    x.At(j) = std::abs(ret.At(j));
```
Comment (pierre-wilmot-fetch-ai) [inline libs/math/include/math/free_functions/standard_functions/abs.hpp]: Need some kind of suppot for fixed point

File: libs/ml/examples/word2vec/main.cpp [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+  std::string  ugly_lookup       = "ugly";
+  SizeType     ugly_idx          = dl.VocabLookup(ugly_lookup);
+  embed_ugly_input->At(ugly_idx) = 1;
+  ArrayType ugly_output          = embeddings->Forward({embed_ugly_input})->Clone();
+
+  ArrayPtrType embed_egg_input = std::make_shared<ArrayType>(dl.VocabSize());
+  std::string  egg_lookup      = "egg";
+  SizeType     egg_idx         = dl.VocabLookup(egg_lookup);
+  embed_egg_input->At(egg_idx) = 1;
+  ArrayType egg_output         = embeddings->Forward({embed_egg_input})->Clone();
```
Comment (pierre-wilmot-fetch-ai) [inline libs/ml/examples/word2vec/main.cpp]: The last 20 lines should probably be collapsed into a function

File: libs/ml/examples/word2vec/main.cpp [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+
+  DataType result_cat_duckling, result_cat_ugly, result_duckling_ugly, result_duckling_egg;
+
+  // distance from cat to duckling (using MSE as distance)
+  result_cat_duckling = fetch::math::distance::Cosine(cat_output, duckling_output);
```
Comment (pierre-wilmot-fetch-ai) [inline libs/ml/examples/word2vec/main.cpp]: Comment doesnt match code (x4)

File: libs/ml/examples/word2vec/main.cpp [hunk 10]
```diff
@@ -0,0 +1,280 @@
...
+    gt->At(0)                          = DataType(input.second);
+    std::shared_ptr<ArrayType> results = g.Evaluate(output_name);
+
+    std::shared_ptr<ArrayType> result =
+        std::make_shared<ArrayType>(std::vector<typename ArrayType::SizeType>({1, 1}));
+
+    DataType tmp_loss = criterion.Forward({results, gt});
+    if (input.second == 0)
+    {
+      tmp_loss *= DataType(p.k_neg_samps);
```
Comment (pierre-wilmot-fetch-ai) [inline libs/ml/examples/word2vec/main.cpp]: Multiplying the loss by 15 ? -- Not sure I understand what's going on here

File: libs/ml/examples/word2vec/main.cpp [hunk 11]
```diff
@@ -0,0 +1,280 @@
...
+    }
+
+    loss += tmp_loss;
+
+    g.BackPropagate(output_name, criterion.Backward({gt, results}));
```
Comment (pierre-wilmot-fetch-ai) [inline libs/ml/examples/word2vec/main.cpp]: criterion.Forward({results, gt});

criterion.Backward({gt, results})

Parameters order is inverted

File: libs/ml/include/ml/graph.hpp [hunk 12]
```diff
@@ -105,23 +105,33 @@ class Graph : public ops::Trainable<T>
...
+        std::make_shared<Node<ArrayType, OperationType>>(name, params...);
     AddNodeImpl<OperationType>(name, inputs, op);
-    FETCH_LOG_INFO("ML_LIB", "Created trainable node [", node_name, "]");
-    trainable_[node_name] = op;
+    FETCH_LOG_INFO("ML_LIB", "Created trainable node [", name, "]");
+    trainable_[name] = op;
     return name;
   }
 
+  std::shared_ptr<fetch::ml::NodeInterface<ArrayType>> GetNode(std::string const &node_name)
```
Comment (pierre-wilmot-fetch-ai) [inline libs/ml/include/ml/graph.hpp]: Should be const

File: libs/ml/include/ml/layers/skip_gram.hpp [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+  {
+
+    // define input and context placeholders
+    std::string input =
+        this->template AddNode<fetch::ml::ops::PlaceHolder<ArrayType>>(name + "_Input", {});
+    std::string context =
+        this->template AddNode<fetch::ml::ops::PlaceHolder<ArrayType>>(name + "_Context", {});
+
+    // embed both inputs
+    embed_in_ = this->template AddNode<fetch::ml::ops::Embeddings<ArrayType>>(
```
Comment (pierre-wilmot-fetch-ai) [inline libs/ml/include/ml/layers/skip_gram.hpp:64]: Should be enough to instanciate only one embeding module

File: libs/ml/include/ml/ops/loss_functions/softmax_cross_entropy.hpp [hunk 14]
```diff
@@ -53,7 +53,8 @@ class SoftmaxCrossEntropy
     assert(inputs.size() == 2);
     assert(inputs[0]->size() == inputs[1]->size());
 
-    typename ArrayType::Type n_classes = static_cast<typename ArrayType::Type>(inputs[1]->size());
+    //    typename ArrayType::Type n_classes = static_cast<typename
```
Comment (pierre-wilmot-fetch-ai) [inline libs/ml/include/ml/ops/loss_functions/softmax_cross_entropy.hpp]: Debug code

File: libs/ml/tests/ml/ops/embeddings.cpp [hunk 15]
```diff
@@ -90,24 +91,31 @@ TYPED_TEST(EmbeddingsTest, backward)
...
-  for (unsigned int j(0); j < 12; ++j)
+
+  for (unsigned int j(0); j < 6; ++j)
+  {
+    errorSignal->Set({0, j}, typename TypeParam::Type(j));
+  }
+  for (unsigned int j(0); j < 6; ++j)
   {
-    errorSignal->Set(j, typename TypeParam::Type(j));
+    errorSignal->Set({1, j}, typename TypeParam::Type(j + 6));
```
Comment (pierre-wilmot-fetch-ai) [inline libs/ml/tests/ml/ops/embeddings.cpp]: These two loops can be collapsed

**Review by pierre-wilmot-fetch-ai — CHANGES_REQUESTED** [review]

**Review by pierre-wilmot-fetch-ai — COMMENTED** [review]
Did a quick review of the C++.
When it comes to the actual Word2Vec thing, I can't really do a proper review since I'm not super familiar with the system.

My intuition is telling me that the curent version of the code is more complex than it needs to be.
I think we should do a proper training on a large enough corpus to make sure it works.
Don't really want to push that kind of code to the public fetch repo before making sure it works.

File: libs/ml/include/ml/dataloaders/w2v_dataloader.hpp [hunk 16]
```diff
@@ -0,0 +1,622 @@
...
+  // training data parsing containers
+  SizeType                                  size_     = 0;     // # training pairs
+  SizeType                                  pos_size_ = 0;     // # positive training pairs
+  SizeType                                  neg_size_ = 0;     // # negative training pairs
+  SizeType                                  n_words_  = 0;     // # of total words in training data
+  std::unordered_map<std::string, SizeType> vocab_;            // full unique vocab
+  std::unordered_map<SizeType, std::string> reverse_vocab_;    // full unique vocab
+  std::unordered_map<std::string, SizeType> vocab_frequency_;  // word frequencies
+  std::vector<double>                       adj_vocab_frequency_;  // word frequencies
+  std::vector<std::vector<std::string>>     words_;                // all training data by sentence
```
Comment (pierre-wilmot-fetch-ai) [inline libs/ml/include/ml/dataloaders/w2v_dataloader.hpp]: Words should be stored as integer, not as strings

**Comment by pierre-wilmot-fetch-ai** [comment]
Playing around with the loader to get an understanding of it.
Got it to crash with that bit of code

```
TEST(W2VDataloaderTest, strip_punctuation_test)
{
  std::string s("Hello, World! My name is FetchBot, I am 1 year old. I eat tokens for breakfast.");
  fetch::ml::W2VLoader<fetch::math::Tensor<float>> loader(s, false, 4, 1, 1, 0);
  auto sample = loader.GetRandom();
}
```
```[1]    7023 floating point exception  ./libs/ml/tests/ml_dataloader_gtest```
**EDIT: FIXED -- My discard threshold was set to 0, which means discard all.**

I'm also not sure what the unigram_table_ role is, but whatever it is, it seems to be doing it in a very inneficient manner, it's allocating a vector of 9999996 elements that all contain.a value bwtween 1 and 15, with all the 1 first, then all the 2, then all the 3, ...

File: libs/math/include/math/free_functions/matrix_operations/matrix_operations.hpp (hunk 3 above)
Comment (pierre-wilmot-fetch-ai) [reply libs/math/include/math/free_functions/matrix_operations/matrix_operations.hpp]: + cur_maxval needs to be updated here !

**Review by pierre-wilmot-fetch-ai — CHANGES_REQUESTED** [review]

File: libs/ml/include/ml/dataloaders/w2v_dataloader.hpp [hunk 17]
```diff
@@ -0,0 +1,622 @@
...
+
+  std::pair<std::pair<std::shared_ptr<T>, std::shared_ptr<T>>, SizeType> GetRandom()
+  {
+    SizeType tmp = static_cast<SizeType>(lcg_());
+    return GetAtIndex(tmp % Size());
+  }
+
+  std::string VocabLookup(SizeType idx)
+  {
+    return vocab_[idx];
```
Comment (pierre-wilmot-fetch-ai) [inline libs/ml/include/ml/dataloaders/w2v_dataloader.hpp]: return reverse_vocab_[idx];

**Review by pierre-wilmot-fetch-ai — CHANGES_REQUESTED** [review]

**Comment by pierre-wilmot-fetch-ai** [comment]
Using new commit, trying to load a lot of data (2450 books from gutenberg project in txt format): 

```
Done loading
Corpus lenght : 51694076
[1]    10101 killed     ./libs/ml/tests/ml_dataloader_gtest
Time: 0h:02m:33s
```

```
Sentences : 454145. Words : 8778897
VocabSize : 205767
```

File: libs/math/tests/math/distance/gtest/cosine_tests.cpp [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include <gtest/gtest.h>
+#include <iomanip>
+#include <iostream>
+
+#include "core/random/lcg.hpp"
```
Comment (pierre-wilmot-fetch-ai) [inline libs/math/tests/math/distance/gtest/cosine_tests.cpp]: Unused

**Review by pierre-wilmot-fetch-ai — CHANGES_REQUESTED** [review]

File: libs/ml/examples/word2vec/main.cpp (hunk 10 above)
Comment (baykaner) [reply libs/ml/examples/word2vec/main.cpp]: was just experimenting with weighting negative samples more heavily; removed

File: libs/math/include/math/free_functions/exponentiation/exponentiation.hpp [hunk 19]
```diff
@@ -37,15 +37,23 @@ namespace fetch {
...
 
 namespace details {
+
 template <typename ArrayType>
-void SquareImpl(ArrayType const &x, ArrayType &ret)
+fetch::math::meta::IfIsMathArray<ArrayType, void> SquareImpl(ArrayType const &x, ArrayType &ret)
 {
   assert(x.size() == ret.size());
-  for (std::size_t i = 0; i < x.size(); ++i)
+  for (typename ArrayType::SizeType i = 0; i < x.size(); ++i)
```
Comment (pierre-wilmot-fetch-ai) [inline libs/math/include/math/free_functions/exponentiation/exponentiation.hpp]: Make that one range loop

File: libs/math/include/math/free_functions/standard_functions/abs.hpp [hunk 20]
```diff
@@ -29,26 +29,61 @@ namespace fetch {
...
+  ret.data().in_parallel().Apply(kernel, x.data());
+}
+
+template <typename ArrayType>
+fetch::math::meta::IfIsNonBlasArray<ArrayType, void> Abs(ArrayType &x, ArrayType &ret)
+{
+  assert(x.size() == ret.size());
+  for (std::size_t j = 0; j < ret.size(); ++j)
+  {
+    Abs(ret.At(j), x.At(j));
```
Comment (pierre-wilmot-fetch-ai) [inline libs/math/include/math/free_functions/standard_functions/abs.hpp]: Parameters seems to be inverted

File: libs/math/include/math/free_functions/standard_functions/abs.hpp [hunk 21]
```diff
@@ -29,26 +29,61 @@ namespace fetch {
...
-fetch::math::meta::IfIsMathArray<ArrayType, void> Abs(ArrayType &x)
+fetch::math::meta::IfIsBlasArray<ArrayType, void> Abs(ArrayType &x, ArrayType &ret)
 {
   math::free_functions::kernels::Abs<typename ArrayType::Type> kernel;
-  x.data().in_parallel().Apply(kernel, x.data());
+  ret.data().in_parallel().Apply(kernel, x.data());
+}
+
+template <typename ArrayType>
+fetch::math::meta::IfIsNonBlasArray<ArrayType, void> Abs(ArrayType &x, ArrayType &ret)
```
Comment (pierre-wilmot-fetch-ai) [inline libs/math/include/math/free_functions/standard_functions/abs.hpp]: Input should be marked const

File: libs/ml/examples/word2vec/main.cpp [hunk 22]
```diff
@@ -0,0 +1,287 @@
...
+{
+  g.AddNode<fetch::ml::ops::PlaceHolder<ArrayType>>("Input", {});
+  g.AddNode<fetch::ml::ops::PlaceHolder<ArrayType>>("Context", {});
+  std::string ret_name = g.AddNode<fetch::ml::layers::SkipGram<ArrayType>>(
+      "SkipGram", {"Input", "Context"}, SizeType(1), SizeType(1), embeddings_size, vocab_size);
+
+  return ret_name;
+}
+
+std::vector<DataType> TestEmbeddings(Graph<ArrayType> &g, std::string &skip_gram_name,
```
Comment (pierre-wilmot-fetch-ai) [inline libs/ml/examples/word2vec/main.cpp]: All input parameters should be marked const

File: libs/ml/examples/word2vec/main.cpp [hunk 23]
```diff
@@ -0,0 +1,287 @@
...
+  std::string husband_lookup = "husband";
+  SizeType    husband_idx    = dl.VocabLookup(husband_lookup);
+  embed_husband_input.At(0)  = DataType(husband_idx);
+  ArrayType husband_output   = embeddings->Forward({embed_husband_input}).Clone();
+
+  ArrayType   embed_wife_input(1);
+  std::string wife_lookup = "wife";
+  SizeType    wife_idx    = dl.VocabLookup(wife_lookup);
+  embed_wife_input.At(0)  = DataType(wife_idx);
+  ArrayType wife_output   = embeddings->Forward({embed_wife_input}).Clone();
```
Comment (pierre-wilmot-fetch-ai) [inline libs/ml/examples/word2vec/main.cpp:120]: Should colapse all these into a GetVector function

File: libs/ml/examples/word2vec/main.cpp [hunk 24]
```diff
@@ -0,0 +1,287 @@
...
+    ArrayType results = g.Evaluate(output_name);
+
+    int pos_count = 0;
+    int neg_count = 0;
+    for (std::size_t j = 0; j < tp.batch_size; ++j)
+    {
+      // result interpreted as probability True - so reverse for gt == 0
+      if (gt.At(j) == DataType(0))
+      {
+        results.At(j)      = DataType(1) - results.At(j);
```
Comment (pierre-wilmot-fetch-ai) [inline libs/ml/examples/word2vec/main.cpp:231]: I dont't understand what's going on here

File: libs/ml/include/ml/dataloaders/skipgram_dataloader.hpp [hunk 25]
```diff
@@ -0,0 +1,281 @@
...
+    double positive_threshold;
+    if (p_.k_negative_samples > 0)
+    {
+      positive_threshold = 1.0 / double(p_.k_negative_samples);
+    }
+    else
+    {
+      positive_threshold = 1.0;
+    }
+    std::vector<SizeType> ret{};
```
Comment (pierre-wilmot-fetch-ai) [inline libs/ml/include/ml/dataloaders/skipgram_dataloader.hpp]: Unused

File: libs/ml/include/ml/dataloaders/skipgram_dataloader.hpp [hunk 26]
```diff
@@ -0,0 +1,281 @@
...
+  /**
+   * randomly select whether to return a positive or negative example
+   */
+  bool SelectValence()
+  {
+    double cur_val = this->lfg_.AsDouble();
+    double positive_threshold;
+    if (p_.k_negative_samples > 0)
+    {
+      positive_threshold = 1.0 / double(p_.k_negative_samples);
```
Comment (pierre-wilmot-fetch-ai) [inline libs/ml/include/ml/dataloaders/skipgram_dataloader.hpp]: This is recomputed at every call to GetRAndom even though it is always yeald the same results

File: libs/ml/include/ml/dataloaders/skipgram_dataloader.hpp [hunk 27]
```diff
@@ -0,0 +1,281 @@
...
+      positive_threshold = 1.0;
+    }
+    std::vector<SizeType> ret{};
+
+    if (cur_val < positive_threshold)
+    {
+      return true;
+    }
+    return false;
+  }
```
Comment (pierre-wilmot-fetch-ai) [inline libs/ml/include/ml/dataloaders/skipgram_dataloader.hpp:121]: return (cur_val < positive_threshold);

File: libs/ml/include/ml/ops/activations/sigmoid.hpp [hunk 28]
```diff
@@ -53,6 +53,12 @@ class Sigmoid : public fetch::ml::ElementWiseOps<T>
       ArrayType const &                                           errorSignal)
   {
     assert(inputs.size() == 1);
+
+    //    std::cout << "inputs.front().get().shape()[0]: " << inputs.front().get().shape()[0] <<
```
Comment (pierre-wilmot-fetch-ai) [inline libs/ml/include/ml/ops/activations/sigmoid.hpp]: Debug code

File: libs/ml/tests/ml/dataloader/text_loader.cpp [hunk 29]
```diff
@@ -0,0 +1,131 @@
...
+  std::vector<std::string> gt_input(
+      {"this", "is", "a", "test", "sentence", "of", "total", "length", "ten", "words",
+       "this", "is", "a", "test", "sentence", "of", "total", "length", "ten", "words"});
+
+  std::string cur_word;
+  for (std::size_t j = 0; j < 20; ++j)
+  {
+    std::pair<TypeParam, SizeType> output = loader.GetNext();
+    cur_word = loader.VocabLookup(SizeType(double(output.first.At(0))));
+    ASSERT_TRUE(cur_word.compare(gt_input.at(j)) == 0);
```
Comment (pierre-wilmot-fetch-ai) [inline libs/ml/tests/ml/dataloader/text_loader.cpp]: ASSERT_EQ(cur_word, gt_input.at(j)) would be more readable

File: libs/math/include/math/free_functions/standard_functions/abs.hpp [hunk 30]
```diff
@@ -29,26 +29,61 @@ namespace fetch {
...
+}
+template <typename Type>
+fetch::math::meta::IfIsArithmetic<Type, Type> Abs(Type &n)
+{
+  Type ret;
+  Abs(n, ret);
+  return ret;
+}
+
+//
```
Comment (pierre-wilmot-fetch-ai) [inline libs/math/include/math/free_functions/standard_functions/abs.hpp]: Dead code

File: libs/ml/include/ml/dataloaders/skipgram_dataloader.hpp (hunk 26 above)
Comment (pierre-wilmot-fetch-ai) [reply libs/ml/include/ml/dataloaders/skipgram_dataloader.hpp]: + by default p_.k_negative_samples is set to 1 (main.cpp:66)
positive_threshold will always be one => Probability of sampling a negitive pair is virtualy 0

**Review by pierre-wilmot-fetch-ai — CHANGES_REQUESTED** [review]

File: libs/ml/include/ml/dataloaders/skipgram_dataloader.hpp (hunk 26 above)
Comment (baykaner) [reply libs/ml/include/ml/dataloaders/skipgram_dataloader.hpp]: good point - i missed a + 1

**Pull request closed without merging**